pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -0,0 +1,146 @@
+type DataplaneCallbacks interface {
+	// OnProxyConnected is executed when proxy is connected after it was disconnected before.
+	OnProxyConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+	// OnProxyReconnected is executed when proxy is already connected, but there is another stream.
+	// This can happen when there is a delay with closing the old connection from the proxy to the control plane.
+	OnProxyReconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+	// OnProxyDisconnected is executed only when the last stream of the proxy disconnects.
+	OnProxyDisconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey)
+}
+
+type xdsCallbacks struct {
+	callbacks DataplaneCallbacks
+	util_xds.NoopCallbacks
+
+	sync.RWMutex
+	dpStreams     map[core_xds.StreamID]dpStream
+	activeStreams map[core_model.ResourceKey]int
+}
+
+func DataplaneCallbacksToXdsCallbacks(callbacks DataplaneCallbacks) util_xds.Callbacks {
+	return &xdsCallbacks{
+		callbacks:     callbacks,
+		dpStreams:     map[core_xds.StreamID]dpStream{},
+		activeStreams: map[core_model.ResourceKey]int{},
+	}
+}
+
+type dpStream struct {
+	dp  *core_model.ResourceKey
+	ctx context.Context
+}
+
+var _ util_xds.Callbacks = &xdsCallbacks{}
+
+func (d *xdsCallbacks) OnStreamClosed(streamID core_xds.StreamID) {
+	d.Lock()
+	defer d.Unlock()
+	dpStream := d.dpStreams[streamID]
+	if dpKey := dpStream.dp; dpKey != nil {
+		d.activeStreams[*dpKey]--
+		if d.activeStreams[*dpKey] == 0 {
+			d.callbacks.OnProxyDisconnected(streamID, *dpKey)
+			delete(d.activeStreams, *dpKey)
+		}
+	}
+	delete(d.dpStreams, streamID)
+}
+
+func (d *xdsCallbacks) OnStreamRequest(streamID core_xds.StreamID, request util_xds.DiscoveryRequest) error {
+	if request.NodeId() == "" {
+		// from https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning:
+		// Only the first request on a stream is guaranteed to carry the node identifier.
+		// The subsequent discovery requests on the same stream may carry an empty node identifier.
+		// This holds true regardless of the acceptance of the discovery responses on the same stream.
+		// The node identifier should always be identical if present more than once on the stream.
+		// It is sufficient to only check the first message for the node identifier as a result.
+		return nil
+	}
+
+	d.RLock()
+	alreadyProcessed := d.dpStreams[streamID].dp != nil
+	d.RUnlock()
+	if alreadyProcessed {
+		return nil
+	}
+
+	proxyId, err := core_xds.ParseProxyIdFromString(request.NodeId())
+	if err != nil {
+		return errors.Wrap(err, "invalid node ID")
+	}
+	dpKey := proxyId.ToResourceKey()
+	metadata := core_xds.DataplaneMetadataFromXdsMetadata(request.Metadata())
+	if metadata == nil {
+		return errors.New("metadata in xDS Node cannot be nil")
+	}
+
+	d.Lock()
+	dpStream := d.dpStreams[streamID]
```

Comment:
Not sure whether streamID semantics allow this to happen, but if a stream opens 2 concurrent requests, then we could increment the active streams once too many. We can mitigate this by checking "d.dpStreams[streamID].dp != nil" again once we have the write lock.